src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs

```diff
@@ -0,0 +1,111 @@
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Builder
+{
+    internal static class AutoRebuildExtensions
+    {
+        // Note that we don't need to watch typical static-file extensions (.css, .js, etc.)
+        // because anything in wwwroot is just served directly from disk on each reload. But
+        // as a special case, we do watch index.html because it needs compilation.
+        // TODO: Make the set of extensions and exclusions configurable in csproj
+        private static string[] _includedSuffixes = new[] { ".cs", ".cshtml", "index.html" };
+        private static string[] _excludedDirectories = new[] { "obj", "bin" };
+
+        public static void UseAutoRebuild(this IApplicationBuilder appBuilder, BlazorConfig config)
+        {
+            // Currently this only supports VS for Windows. Later on we can add
+            // an IRebuildService implementation for VS for Mac, etc.
+            if (!VSForWindowsRebuildService.TryCreate(out var rebuildService))
+            {
+                return; // You're not on Windows, or you didn't launch this process from VS
+            }
+
+            var currentCompilationTask = Task.CompletedTask;
+            var compilationTaskMustBeRefreshedIfNotBuiltSince = (DateTime?)null;
+
+            WatchFileSystem(config, () =>
+            {
+                // Don't start the recompilation immediately. We only start it when the next
+                // HTTP request arrives, because it's annoying if the IDE is constantly rebuilding
+                // when you're making changes to multiple files and aren't ready to reload
+                // in the browser yet.
+                compilationTaskMustBeRefreshedIfNotBuiltSince = DateTime.Now;
+            });
+
+            appBuilder.Use(async (context, next) =>
+            {
+                try
+                {
+                    if (compilationTaskMustBeRefreshedIfNotBuiltSince.HasValue)
+                    {
+                        currentCompilationTask = rebuildService.PerformRebuildAsync(
+                            config.SourceMSBuildPath,
+                            compilationTaskMustBeRefreshedIfNotBuiltSince.Value);
+                    }
+
+                    await currentCompilationTask;
+                }
+                catch (Exception)
+                {
+                    // If there's no listener on the other end of the pipe, or if anything
+                    // else goes wrong, we just let the incoming request continue.
+                    // There's nowhere useful to log this information so if people report
+                    // problems we'll just have to get a repro and debug it.
+                    // If it was an error on the VS side, it logs to the output window.
+                }
+                finally
+                {
+                    compilationTaskMustBeRefreshedIfNotBuiltSince = null;
+                }
```

Comment:
How concerned are you about thread safety of this interchange. In particular if another thread observes a file change during a build, it looks like this would wipe out the `compilationTaskMustBeRefreshedIfNotBuiltSince`, which would have been set to a newer value than you just built with

Reply:
Good point. I'll change the ordering of operations to avoid that issue.